server.lisp

```diff
@@ -184,3 +186,59 @@
   (server-send-message call bytes-to-send)
   (server-send-status call)
   (server-recv-close call))
+
+(defun run-methods (methods server)
+  (let* ((call (start-call-on-server server))
+         (messages (receive-message-from-client call))
+         (message (apply #'concatenate '(array (unsigned-byte 8) (*)) messages)))
+    (unwind-protect
+      (let* ((method (find (call-method-name call)
+                           methods
+                           :test
+                           #'string=
+                           :key #'method-details-name))
+             (deserialized-message (funcall (method-details-deserializer method)
+                                            message))
+             (response (funcall (method-details-action method)
+                                deserialized-message))
+             (serialized-response (funcall (method-details-serializer method)
+                                           response)))
+        (format t "~%~A~%" response)
+        (send-message-to-client call serialized-response))
+      (free-call-data call))))
+
+(defun run-grpc-server (address methods
+                        &key
+                        (server-creds
+                         (grpc-insecure-server-credentials-create))
+                        (cq grpc::*completion-queue*)
+                        (num-threads 1)
+                        (run-methods #'run-methods)
+                        (exit-count nil))
+  (let* ((server (start-server cq server-creds address))
+         (sem (bordeaux-threads:make-semaphore :count num-threads))
+         threads)
+
+    (dolist (method methods)
+      (format t "~s~%" (method-details-name method))
+      (register-method server (method-details-name method) address))
+    (run-server server server-creds)
+
+    (unwind-protect
+         (loop for run-count from 0
+               while (or (not exit-count)
+                         (< run-count exit-count))
+               do
+            (bordeaux-threads:wait-on-semaphore sem)
+            (push
+             (bordeaux-threads:make-thread
+              (lambda ()
+                (funcall run-methods methods server)
+                (bordeaux-threads:signal-semaphore sem))
+              :name (format nil "Run Method Run ~a" run-count))
+             threads)))
```

Comment:
It's late and I'm tired, but if exit-count is nil won't this push a potentially infinite number of threads onto `threads`?

Reply:
It wouldn't, but I pushed that down to dispatch-requests.